torch/csrc/utils/tensor_new.cpp

```diff
@@ -148,6 +148,15 @@ std::vector<int64_t> compute_sizes(PyObject* seq) {
 }
 
 ScalarType infer_scalar_type(PyObject *obj) {
+#ifdef USE_NUMPY
```

Comment:
Why did we move this code block?

Reply:
Functions below like THPUtils_checkLong() are checking if it's a numpy long or python long. If we run the numpy checks last, we'll never reach the numpy checks so won't check the numpy dtype.

Reply:
If `THPUtils_checkLong(obj)` decides that obj is a numpy long, then `infer_scalar_type` returns `ScalarType::Long`. This seems correct to me -- what is wrong there?

Reply:
PyArray_CheckScalar will return true for a np.int32 scalar, and get the correct torch.int32 result type. If we rely on THPUtils_checkLong we always return Long.

Reply:
I'm confused. The test https://github.com/pytorch/pytorch/pull/30486/files/12d6de600476b9b1d9bcae6351680b37995e6d9c#diff-f539198dd366265fb8dc2d661bc5d5bcL1507 seems to test that a np.int32 gets converted to a torch.IntTensor, which worked before this PR. Is the test wrong?

Reply:
Yes the test was semi-broken on master (and fixed by this PR).

On master, THPUtils_checkDouble checks for numpy scalars, but THPUtils_checkLong doesn't. This PR changes checkLong  to also check for numpy ints, consistent with checkDouble. On master the test was broken for float/double depending on the result of get_default_scalar_type() below. (the actual issue on master was a short-circuiting for np.float/np.double not for int32, though the same issue gets exposed for int32 as well once changing other things in this PR).

See here for a related change:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Reply:
Pitch: Instead of moving the codeblock, we should change the call to `THPUtils_checkLong` below to a call to a new function `THPUtils_checkPythonLong` (or introduce a `THPUtils_checkLong(PyObject*, bool include_numpy_long=True)`)

Before this pull request, all of the cases are exclusive; none of them overlap. After this pull request, some of the checks overlap. This is bad for code readability: "Does THPUtils_checkLong mean check that the value is convertible to a long, or is it checking for a python long?"

Making the checks exclusive also means that we don't have to worry about the order in which they appear, like we are doing now. With this pull request, there is a non-trivial reason things in this function are ordered the way they are.

Reply:
> Before this pull request, all of the cases are exclusive

I don't think that's true. Before this PR, a numpy float scalar was being caught by one these checks and short circuiting to return `get_default_dtype` without reaching the numpy handling block (which it also would have satisfied). To make this mutually exclusive, we'd also need something like maybe a wrapper function to check `PyFloat_Check and not numpy`. To me, that seems like overkill, at least for purposes of this bug-fix.

We can keep iterating after this lands too if we can come up with a better code structure.

Reply:
Which case does a `numpy.float64` scalar go into?

Reply:
Oh I see, PyFloat_check returns True for numpy.float64... On the other hand, PyLong_check returns False for numpy.int64 :/

We could use PyFloat_CheckExact instead of PyFloat_Check. But if you think this is overkill, we should at least include a comment in the code explaining that the numpy checks must go first otherwise they'll be considered to act like a python float or long.